### PR TITLE
Home work

### PR DESCRIPTION
**Pull request by ChistEgor**
added 1 task. second is coming soon

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import col, split, row_number, explode
+from pyspark.sql import Window
+
+if __name__ == '__main__':
+    spark = SparkSession \
+        .builder \
+        .master("local[12]") \
```
Comment (wena77) [inline main.py]: you can put `*` here to use all available logical cores 
and pls choose one style: `double` or `single` quotes

File: main.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        .appName("main") \
+        .getOrCreate()
+
+    """
+    The First Task'a
+    """
```
Comment (wena77) [inline main.py]: doc-string should be after func definition, if you want to put comments inside function use `#`

File: main.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    df_tb = spark.read.csv('input_data/title.basics.tsv', sep=r'\t', header=True)
+    df_tr = spark.read.csv('input_data/title.ratings.tsv', sep=r'\t', header=True)
```
Comment (wena77) [inline main.py]: move this functionality to the separate function

File: main.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    df_tb_tr = df_tb.join(df_tr, 'tconst', 'inner')
+
+    all_top_movie = df_tb_tr \
+        .select('tconst', 'primaryTitle', 'numVotes', 'averageRating', 'startYear') \
+        .where((col('titleType') == 'movie') & (col('numVotes') >= 100000)) \
+        .orderBy(col('averageRating').desc()).limit(100)
+    all_top_movie.write.mode('overwrite').save('output_data/1_1', format='csv')
```
Comment (wena77) [inline main.py]: move this functionality to the separate function

File: main.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        .builder \
+        .master("local[12]") \
+        .appName("main") \
+        .getOrCreate()
+
+    """
+    The First Task'a
+    """
+
+    df_tb = spark.read.csv('input_data/title.basics.tsv', sep=r'\t', header=True)
```
Comment (wena77) [inline main.py]: make names of your variables more intuitive

File: main.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    top_movie_in_60s.write.mode('overwrite').save('output_data/1_3', format='csv')
+
+    """
+    The Second Task'a
+    """
+
+    genre_explode = explode(split(df_tb_tr.genres, ',')).alias('genre')
+
+    all_top_movie = df_tb_tr \
+        .select('tconst', 'primaryTitle', 'numVotes', genre_explode, 'averageRating', 'startYear') \
```
Comment (wena77) [inline main.py]: you can do `.withColumn` after select, it's more readable and you won't need `alias` then

File: main.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        .where((col('titleType') == 'movie') & (col('numVotes') >= 100000))
+
+    top_movie_each_genre = all_top_movie \
+        .select('tconst', 'primaryTitle', 'startYear', 'genre', 'averageRating', 'numVotes',
+                row_number().over(Window.partitionBy('genre').orderBy(col('averageRating').desc())).alias(
```
Comment (wena77) [inline main.py]: better do it with `withColumn`
move part with `Window`  to the variable

File: main.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    df_tb_tr = df_tb.join(df_tr, 'tconst', 'inner')
+
+    all_top_movie = df_tb_tr \
+        .select('tconst', 'primaryTitle', 'numVotes', 'averageRating', 'startYear') \
+        .where((col('titleType') == 'movie') & (col('numVotes') >= 100000)) \
+        .orderBy(col('averageRating').desc()).limit(100)
+    all_top_movie.write.mode('overwrite').save('output_data/1_1', format='csv')
+
+    top_movie_last_10y = all_top_movie.where(col('startYear') > 2000)
```
Comment (wena77) [inline main.py]: why `2000`?
how will this work in 2035? :D

Comment (ChistEgor) [reply main.py]: Fair enough :@

File: first_task.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+from datetime import datetime
+from pyspark.sql import SparkSession
+from data import info_cinema, info_ratings
+from pyspark.sql.functions import col
```
Comment (wena77) [inline first_task.py]: Imports should be grouped in the following order:

- Standard library imports.
- Related third party imports.
- Local application/library specific imports.

You should put a blank line between each group of imports.

Comment (ChistEgor) [reply first_task.py]: okey

File: functions.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def find_top_movie(info_cinema, info_ratings):
+    """
+    Filters each movie and return the top 100
+    """
+    return info_cinema \
+        .join(info_ratings, 'tconst', 'inner') \
+        .where((col('titleType') == 'movie') & (col('numVotes') >= 100000)) \
+        .orderBy(col('averageRating').desc()).limit(100)
```
Comment (wena77) [inline functions.py]: don't do `limit` here

File: functions.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def filter_top_movie(df):
+    """
+    Only films from 1949 until now
+    """
+    return df.where(col('startYear') > 1949)
+
+
+def make_year_range(df):
```
Comment (wena77) [inline functions.py:58]: you use it only once, do we really need it as a separate function?

File: data.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+info_names = functions.read_data('name.basics.tsv')
+info_cinema = functions.read_data('title.basics.tsv')
+info_ratings = functions.read_data('title.ratings.tsv')
+info_staff = functions.read_data('title.principals.tsv')
+info_crew = functions.read_data('title.crew.tsv')
+
+# For 1-5
+top_all_time = functions.find_top_movie(info_cinema, info_ratings)
+# For 3
+top_from_1950 = functions.filter_top_movie(top_all_time)
```
Comment (wena77) [inline data.py]: get rid of this file, move it to the `main`

Comment (ChistEgor) [reply data.py]: okey

File: main.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import functions
+
+info_names = functions.read_data('name.basics.tsv')
+info_cinema = functions.read_data('title.basics.tsv')
+info_ratings = functions.read_data('title.ratings.tsv')
+info_staff = functions.read_data('title.principals.tsv')
+info_crew = functions.read_data('title.crew.tsv')
+
+# For 1-5
+top_all_time = functions.find_top_movie(info_cinema, info_ratings)
```
Comment (wena77) [inline main.py]: move to `__main__`

File: main.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import tasks
+import functions
```
Comment (wena77) [inline main.py]: import all you need at once

File: functions.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    Reads csv file from input_data
+    """
+    return spark.read.csv(f'input_data/{file_name}', sep=r'\t', header=True, inferSchema=True)
+
+
+def write_to_csv(data, file):
+    """
+    Makes one dataframe from group of and saves to file
+    """
+    return data.coalesce(1).write.mode('overwrite').save(f'output_data/{file}', format='csv')
```
Comment (wena77) [inline functions.py]: please add `headers` to output files

Comment (ChistEgor) [reply functions.py]: Okey

**Comment by ChistEgor** [comment]
🐱‍👤

**Pull request merged**